sending/backends/websocket.py

```diff
@@ -0,0 +1,205 @@
+        """
+        Websocket-based Sending backend. This class handles creating the initial
+        websocket connection and reconnecting on server-side disconnect.
+
+        One key difference between this Backend and others is that it ignores
+        the concept of `topic`. All messages seen over the wire are dropped
+        onto the inbound queue with topic '', and .send() is overwritten
+        to only take a message, defaulting outbound topic to ''.
+
+        Real-world applications are likely to subclass this and add hooks
+        either as class methods or attach them to an instance after init.
+
+         - inbound_message_hook: deserialize incoming messages over websocket
+         - outbound_message_hook: serialize outgoing messages over websocket
+         - init_hook: called after websocket connection is established
+         - auth_hook: called just before init_hook, useful if you need to send
+                      some kind of auth request. This effects .send(), make sure
+                      you also register a callback to .on_auth that is called
+                      when you receive an auth response.
+        """
+        super().__init__()
+        self.ws_url = ws_url
+
+        # An unauth_ws and authed_ws pair of Futures are created so that
+        # sub-classes can easily implement a pattern where messages are only
+        # sent to the server after the session has been authenticated.
+        self.unauth_ws = asyncio.Future()
+        self.authed_ws = asyncio.Future()
+
+        # When an outbound worker is ready to send a message over the wire, it
+        # calls ._publish which will await the unauth_ws or authed_ws Future.
+        # However, if something goes wrong and those Futures never resolve then
+        # it can leave you in a hard-to-debug state. So ._publish should use an
+        # asyncio.wait_for(await <ws future>, timeout=self.publish_timeout)
+        self.publish_timeout: float = 5.0
+
+        # Used to prevent automatic websocket reconnect when we're trying to shutdown
+        self._shutting_down = False
+
+        # For debug and testing, .next_event gets set/cleared after every received message
+        self.next_event = asyncio.Event()
+        self.last_seen_message = None
+        self.reconnections = 0
+        # If this is set, then the WebsocketManager will stop reconnecting after this
+        # many reconnections.
+        self.max_reconnections = None
+
+        # Optional hooks that can be defined in a subclass or attached to an instance.
+        # First auth_hook then init_hook are called immediately after websocket connection
+        # They're separated into two hooks in case a subclass needs to authenticate, and
+        # also queue up initial messages that use .send() (which will wait until auth is accepted)
+        if not hasattr(self, "auth_hook"):
+            self.auth_hook: Optional[Callable] = None
+        self.init_hook: Optional[Callable] = None
```

Comment:
At some point I think we could graduate more hooks to the base manager. Could see some backends benefitting from the ability to define their own auth hook or init hook.